### DISCUSSION TITLE
如何自定义编译

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#1664] [Q&A] [answered]
您好，我想咨询1：如果我想根据某个.h中是否#define xxx 或者#define xxx的值来判断是否编译某些.c文件该如何操作？
2,：如果使用xmake传入自定义参数，根据这些来判断是否编译某些.c文件，类似上面该如何操作？谢谢

**Comment by waruqi** [comment]
> 您好，我想咨询1：如果我想根据某个.h中是否#define xxx 或者#define xxx的值来判断是否编译某些.c文件该如何操作？

自己走 #if #include "xx.c" 这跟 xmake 没关系

> 2,：如果使用xmake传入自定义参数，根据这些来判断是否编译某些.c文件，类似上面该如何操作？谢谢

option("xxx") + has_config("xxx")

**Comment by PeakRacing** [comment]
主要是想获取.h中的一些#define信息来判断决定xmake是否加入某些target，还有after_build中的一些差异的打包行为

**Comment by waruqi** [comment, marked as answer]
定义 target 属于描述阶段就定义好的，后期读文件判断就置后了。。没法再调整，只能在 on_load() 阶段，动态通过 target:set("default", false) 来禁用某个 target 的默认编译，你也可以在 on_load 阶段走 io.readfile 读取指定文件 判断 defines

**Comment by PeakRacing** [comment]
了解，非常感谢
